TESTING.md

```diff
@@ -0,0 +1,110 @@
+# Testing
+
+Tools used to test the application:
+
+* **ESLint** (static code analysis of JS files)
+* **Markdownlint** (static analysis of MD files)
+* **Stylelint** (static code analysis of CSS files)
+* **Jest** (unit testing)
+* **Playwright** (visual and functional component testing)
+
+Generally, `npm test` and `npm run test:playwright-ct:all` should be run  within
+their designated Docker containers before pushing changes to the repository.
+
+## Tools
+
+### Linters (ESlint, Markdownlint, Stylelint)
+
+They can be run individually:
+
+1. (optional) [Open][gh-gg-node-shell] `node_shell` Docker
+   container
+2. Run linters:
+
+    ```bash
+    npm run <eslint|markdownlint|stylelint>
+    ```
+
+or all together:
+
+1. (optional) [Open][gh-gg-node-shell] `node_shell` Docker
+   container
+2. Run linters:
+
+    ```bash
+    npm run lint
+    ```
+
+### Jest
+
+Jest tests can be run using the following commands:
+
+1. (optional) [Open][gh-gg-node-shell] `node_shell` Docker
+   container
+2. Run tests:
+
+     ```bash
+     npm run test:jest
+     ```
+
+or all together with linters:
+
+1. (optional) [Open](/README.md#node_shell-docker-container) `node_shell` Docker
+   container
+2. Run tests:
+
+    ```bash
+    npm test
+    ```
+
+### Playwright
+
+Playwright tests must be run in a Docker container to ensure a uniform
+environment. Otherwise, snapshots would differ between operating systems.
+
+This is the reason why you need to run Playwright tests separately
+from other tools.
+
+#### Configuration
+
+Test parameters can be tweaked by creating and tweaking `.env.playwright` file:
+
+```bash
+cp .env.playwright.dist .env.playwright
+```
+
+#### Running Tests
+
+Playwright tests can be run using the following commands:
+
+1. [Open][gh-gg-playwright] `playwright` Docker container
+2. Run tests:
+
+    ```bash
+    npm run test:playwright-ct:<all|all-with-update>
+    ```
+
+You can also run specific tests by passing a path to the test files:
+
+```bash
+npm run test:playwright-ct:<all|all-with-update> -- <match_path>
+```
+
+You can also pass any [CLI command][playwright-cli] to the test runner:
+
+```bash
+npm run test:playwright-ct:<all|all-with-update> -- <cli_argument>
+```
+
+#### Opening Test Report
+
+After running Playwright tests, test report can be served into a browser
+by using the following command:
+
+```bash
+npx playwright show-report
+```
```

Comment:
Does it work when run from inside of the Docker container?

Reply:
Unfortunatelly not. I am glad you catch this as I was originally mentioning host system in the docs, but I must have removed in when changing texts to match styling used in the rest of the documentation.

I made slight improvement so you can change port if you need (this is just part of .env.playwright as always). But it is not exposible from the Docker container yet. I have tried several things to expose it but without success. But I do not see problem running report from the host system or opening HTML file directly. 

Even though `npx playwright show-report` is standard approach and recommended by output after Playwrifght finishes, I added `test:playwright-ct:show-report` script that runs it with variable port.

Reply:
I might probably create issue in Playwright repository when I have time .

Reply:
Honestly I just run `cd playwright-report` and then `python -m http.server 8000`.
It then serves the report on port 8000.

No need to use Playwright for that and it works outside of docker. I think virtually every modern *NIX system has got Python installed.

Reply:
I reacted to another conversation. TLDR; Check latest TESTING.md. Now it can be run from Docker, reason is described at another comment.

Reply:
Just to clarify this. This approach is simply the standard approach recommended by Playwright. Reason is that Playwright can generate not only HTML report, but JSON report and much more. This command always serves the report despite the report type. I might be better to run JSON reports on CI as it is more readable than HTML source, but it is up to another discussion.